src/hal/src/acceleration_structure.rs

```diff
@@ -0,0 +1,224 @@
+        const OPAQUE = 0x1;
+        /// indicates that the implementation must only call the any-hit shader a single time for each primitive in this geometry. If this bit is absent an implementation may invoke the any-hit shader more than once for this geometry.
+        const NO_DUPLICATE_ANY_HIT_INVOCATION = 0x2;
+    }
+}
+
+#[derive(Debug)]
+pub struct AccelerationStructureGeometry<'a, B: Backend> {
+    pub flags: AccelerationStructureGeometryFlags,
+    pub geometry: Geometry<'a, B>,
+}
+
+#[derive(Debug)]
+pub enum Geometry<'a, B: Backend> {
+    Triangles(GeometryTriangles<'a, B>),
+
+    /// TODO bikeshed capitalization of AABBs.
+    Aabbs(GeometryAabbs<'a, B>),
+
+    Instances(GeometryInstances<'a, B>),
+}
+
+// https://www.khronos.org/registry/vulkan/specs/1.2-extensions/man/html/VkAccelerationStructureGeometryTrianglesDataKHR.html
+// for docs: Memory Safety (https://nvpro-samples.github.io/vk_raytracing_tutorial_KHR/#accelerationstructure)
+#[derive(Debug)]
+pub struct GeometryTriangles<'a, B: Backend> {
+    /// Both APIs require support for at least the following:
+    /// - `(R32_G32, Float)`
+    /// - `(R32_G32_B32, Float)`
+    /// - `(R16_G16, Float)`
+    /// - `(R16_G16_B16_A16, Float)`
+    /// - `(R16_G16, Inorm)`
+    /// - `(R16_G16_B16_A16, Inorm)`
+    /// VK could support more by querying `VK_FORMAT_FEATURE_ACCELERATION_STRUCTURE_VERTEX_BUFFER_BIT_KHR`, DX12 is not queryable? Note [the DX12 ray tracing spec](https://microsoft.github.io/DirectX-Specs/d3d/Raytracing.html#d3d12_raytracing_geometry_triangles_desc) says it supports more than [the Win32 docs](https://docs.microsoft.com/en-us/windows/win32/api/d3d12/ns-d3d12-d3d12_raytracing_geometry_triangles_desc).
+    pub vertex_format: Format,
+    pub vertex_buffer: &'a B::Buffer,
+    pub vertex_buffer_offset: Offset,
+    pub vertex_buffer_stride: Offset,
+
+    /// aka "vertex count"?
+    pub max_vertex: Offset,
+
+    // index format must be DXGI_FORMAT_R32_UINT, DXGI_FORMAT_R16_UINT
+    // can also be VK_INDEX_TYPE_NONE_KHR/DXGI_FORMAT_UNKNOWN
+    pub index_buffer: Option<(&'a B::Buffer, Offset, IndexType)>,
+
+    /// 3x4 matrix
+    /// TODO enum for cpu repr?
+    /// Must point to an address containing `TransformMatrix`
+    pub transform: Option<(&'a B::Buffer, Offset)>,
+}
+
+// VkTransformMatrixKHR
+#[derive(Debug)]
+pub struct TransformMatrix {
+    /// 3x4 row-major affine transformation matrix. Use `mint::RowMatrix3x4` if available.
+    pub matrix: [[f32; 4]; 3],
+}
+
+#[derive(Debug)]
+pub struct GeometryAabbs<'a, B: Backend> {
+    /// Must point to a buffer with buffer with an array of `AabbPositions`s
+    /// TODO: document lifetime required for this buffer
+    pub buffer: &'a B::Buffer,
+    pub buffer_offset: Offset,
+    pub buffer_stride: Offset,
+}
+
+#[derive(Debug)]
+pub struct AabbPositions {
+    /// Use `mint::Vector3` if available.
```

Comment:
we aren't using `mint` for gfx-hal at this point

Reply:
I'll reword all mint references to describe the vector kind